Feature/programming exercise/task specific hints + improved instruction analysis

thilo-behnke opened the pull request:
<!-- Thanks for contributing to ArTEMiS! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
- [x] I tested the changes and all related features on the test server https://artemistest.ase.in.tum.de.
- [x] I documented my source code using the JavaDoc / JSDoc style.
- [x] ~I added integration test cases for the server (Spring) related to the features~
- [x] I added integration test cases for the client (Jest) related to the features
- [x] I added screenshots/screencast of my UI changes
- [x] I translated all the newly inserted strings

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->

Atm exercise hints can only be added the whole programming exercise.
However in many cases it would be more useful to directly assign a hint to a task.

### Description
<!-- Describe your changes in detail -->

This PR will add functionality for assigning hints to tasks of an exercise:
- Domain Command for adding task specific hints
- Auto completion in editor for adding hints
- Tasks will be rendered in HTML **only for failed tests**

Also it will improve the current problem statement analysis that is available in the `programming-exercise-editable-instruction.component`:
- Show warning in line number in editor (see screenshot)

:information_source: **Please read before reviewing:**
- The markers are not working well for, which is why 'only' the line number is marked and not the exact text.
- There is some code that was just moved around in the domain-multi-option command which is not new and does not have tests (we don't have a concept on how to test the markdown-editor, I added this as an action item to our meeting)

**Status:**
- [x] Add domain command for inserting hints
- [x] Integrate hints into task.extension parsing
- [x] Integrate hints into task.component render
- [x] Improve instruction analysis
- [x] Update tests
- [x] Update docs
- [x] Write new tests

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->

**Adding hints to tasks:**

1. Log in to ArTEMiS
2. Navigate to Course Administration
3. Choose edit for a programming exercise that has hints
4. Use the task domain command to add a task
5. Replace the sample hintIds in the task template with a real hint id (e.g. by using the hint domain command)
6. Switch to preview: The hint should now be rendered at the end of the task **if it's failing**

**Analyzing the problem statement:**

1. Log in to ArTEMiS
2. Navigate to Course Administration
3. Choose edit for a programming exercise that has hints & test cases
4. Do something wrong in the problem statement, e.g. add a test case or hint that does not exist
=> There should be a warning on the appropriate line numbers

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->

**Adding task specific hints:**

![image](https://user-images.githubusercontent.com/28230611/63649593-c5635800-c73f-11e9-8d1c-d139cdda2d48.png)

**Rendering in HTML for failed tasks:**

![image](https://user-images.githubusercontent.com/28230611/63649618-0bb8b700-c740-11e9-909b-4fe560eed714.png)

**Line number analysis:**

![image](https://user-images.githubusercontent.com/28230611/63649578-96e57d00-c73f-11e9-9953-1d6aa43d10bd.png)

ungaralex commented on src/main/webapp/i18n/de/programmingExercise.json:
```diff
@@ -63,7 +64,14 @@
+                "invalidTestCase": "Der Testfall '{{id}}' konnte nicht im Test Repository gefunden werden."
+            },
+            "hintsAnalysis": {
+                "invalidHints": "{{number}} Hinweise wurden im Markdown der Aufgabenstellung gefunden, die nicht in der Datenbank existieren bzw. nicht dieser Programmierübung zugeordnet sindÖ",
```
```suggestion
                "invalidHints": "{{number}} Hinweise wurden im Markdown der Aufgabenstellung gefunden, die nicht in der Datenbank existieren bzw. nicht dieser Programmierübung zugeordnet sind",
```

thilo-behnke commented:
> Really great work! I really like the code and the feature itself.
> I only have one small issue: As I am adding hints in the markdown and pressing Strg+Space, or just receiving general suggestions, it would be nice to get hints suggested by their name and not their ID. E.g. for a hint {name="Great Hint", id="12"}, if I start typing "Gre..." I should get the actual hint suggested and not only tasks starting with "Gre...". If this is possible, it would be nice to have it, otherwise I have no complaints.

Thanks for the review! Good catch, I will look into it.

thilo-behnke commented:
> > Really great work! I really like the code and the feature itself.
> > I only have one small issue: As I am adding hints in the markdown and pressing Strg+Space, or just receiving general suggestions, it would be nice to get hints suggested by their name and not their ID. E.g. for a hint {name="Great Hint", id="12"}, if I start typing "Gre..." I should get the actual hint suggested and not only tasks starting with "Gre...". If this is possible, it would be nice to have it, otherwise I have no complaints.
> 
> Thanks for the review! Good catch, I will look into it.

It doesn't seem to be possible, appearently there is no option to search for something else than the value that is inserted (here the id) :( https://github.com/ajaxorg/ace/wiki/How-to-enable-Autocomplete-in-the-Ace-editor

There is however the possibility to press CTRL+Space to see all available completions.

I made some performance improvements to the instruction analysis.

krusche commented on src/main/webapp/app/markdown-editor/markdown-editor.component.ts:
```diff
@@ -76,6 +77,8 @@ export class MarkdownEditorComponent implements AfterViewInit {
     @Input() markdown: string;
     @Input() editorMode = EditorMode.NONE;
+    // If false will hide the line numbers.
+    @Input() showGutter = false;
```
can we rename this input to something like showLineNumbers?

thilo-behnke replied:
You are right, then we also don't need a comment for it.

krusche commented on src/main/webapp/app/entities/programming-exercise/instructions/instructions-editor/analysis/programming-exercise-instruction-instructor-analysis.component.ts:
```diff
@@ -0,0 +1,83 @@
+import { Component, EventEmitter, Input, OnChanges, OnDestroy, OnInit, Output, SimpleChanges } from '@angular/core';
```
I guess you can change the filename to `programming-exercise-instruction-analysis.component.ts`
Same for the other files. I believe that `instruction-instructor` is not necessary and too much

thilo-behnke replied:
You are right, the wording was not good, I changed it.

krusche commented:
I just tested this feature on the test server and it looks good to me.
One remark: Do we really need a Cancel button? I would leave this out and allow the user to simply click somewhere else than the popup to close the dialog.

Or, I would at least rename the button to Close.

<img width="123" alt="image" src="https://user-images.githubusercontent.com/744067/64019438-75b1d180-cb2f-11e9-8a70-06ef7a6029a6.png">

thilo-behnke commented:
> I just tested this feature on the test server and it looks good to me.
> One remark: Do we really need a Cancel button? I would leave this out and allow the user to simply click somewhere else than the popup to close the dialog.
> 
> Or, I would at least rename the button to Close.
> 
> <img alt="image" width="123" src="https://user-images.githubusercontent.com/744067/64019438-75b1d180-cb2f-11e9-8a70-06ef7a6029a6.png">

Thanks for your feedback!
I removed the Cancel button and added the option to close the modal on click in the background.
